.github/workflows/main.yaml

```diff
@@ -183,12 +209,79 @@ jobs:
+          needs.initialize.outputs.on_device_test == 'true' &&
+          (
+            github.event_name != 'pull_request' ||
+            contains(github.event.pull_request.labels.*.name, 'on_device')
+          )
         uses: ./.github/actions/upload_test_artifacts
         with:
           type: ondevice
           os: linux
 
+  # Runs on-host integration and unit tests.
+  on-device-test:
+    needs: [initialize, build]
+    if: |  # Always run if not a PR or if the PR has on_device label
+      needs.initialize.outputs.on_device_test == 'true' && (
+        github.event_name != 'pull_request' ||
+        contains(github.event.pull_request.labels.*.name, 'on_device')
+      )
+    runs-on: [self-hosted, linux, X64]
+    name: on_device_${{ matrix.name }}_${{ matrix.type }}
+    container: ${{ needs.docker-unittest-image.outputs.docker_unittest_tag }}
+    strategy:
+      fail-fast: false
+      matrix:
+        platform: ${{ fromJson(needs.initialize.outputs.platforms) }}
+        config: [devel]
+        type: ${{ fromJson(needs.initialize.outputs.on_device_test_tests) }}
+        include: ${{ fromJson(needs.initialize.outputs.includes) }}
+    permissions:
+      packages: none
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
+      - name: Install requirements
+        run: |
+          pip3 install grpcio==1.37.0 grpcio-tools==1.38.0
+        shell: bash
+      - name: Generate gRPC files
+        run: |
+          python -m grpc_tools.protoc -Itools/ --python_out=tools/ --grpc_python_out=tools/ tools/on_device_tests_gateway.proto
+        shell: bash
+      - name: Set up Cloud SDK
+        uses: google-github-actions/setup-gcloud@v0
+      - name: Set env vars
+        run: |
+          echo "PROJECT_NAME=$(gcloud config get-value project)" >> $GITHUB_ENV
+          echo "GITHUB_RUN_NUMBER=${GITHUB_RUN_NUMBER}" >> $GITHUB_ENV
+          echo "WORKFLOW=${{ github.workflow }}" >> $GITHUB_ENV
+          echo "LOGS_FILE_DIR=$GITHUB_WORKSPACE/out/tmp/${{ matrix.platform }}_${{ matrix.config }}_${{ matrix.type }}/" >> $GITHUB_ENV
+      - name: Create logs directory
+        run: |
+          mkdir -p "${LOGS_FILE_DIR}"
+        shell: bash
+      - name: run ${{ matrix.type }} tests on ${{ matrix.platform }} platform
+        id: on_device_test
+        run: |
+          python3 tools/on_device_tests_gateway_client.py \
+            --token ${{ github.token }} \
+            --test_type ${{ matrix.type }} \
+            --platform ${{ matrix.target_platform }} \
+            --config ${{ matrix.config }} \
+            --label cobalt_github_${{ github.event_name }} \
+            --tag cobalt_github_${{ github.event_name }} \
+            --change_id "${{ github.sha }}" \
+            --builder_name github_${{ matrix.platform }}_tests \
+            --build_number ${GITHUB_RUN_NUMBER} \
+            --loader_config ${{ matrix.config }} \
+            --loader_platform ${{ needs.initialize.outputs.bootloader }} \
```

Comment:
if 'bootloader' is not set in json it will be 'null' here, see https://screenshot.googleplex.com/6mHu4Jdr7SarWP9.png.  We need to handle that.